binding.cc

```diff
@@ -0,0 +1,1773 @@
+#include <bare.h>
+#include <js.h>
+#include <jstl.h>
+#include <stdlib.h>
+#include <string.h>
+#include <udx.h>
+
+#define UDX_NAPI_INTERACTIVE     0
+#define UDX_NAPI_NON_INTERACTIVE 1
+#define UDX_NAPI_FRAMED          2
+
+namespace {
+// socket
+using cb_socket_send_t = js_function_t<void, js_receiver_t, uint64_t, int>;
+using cb_socket_message_t = js_function_t<js_typedarray_span_t<>, js_receiver_t, int64_t, int, std::string, int>;
+using cb_socket_close_t = js_function_t<void, js_receiver_t>;
+using cb_socket_realloc_message_t = js_function_t<js_typedarray_span_t<>, js_receiver_t>;
+
+// stream
+using cb_stream_end_t = js_function_t<void, js_receiver_t, uint32_t>;
+using cb_stream_data_t = js_function_t<js_typedarray_span_t<>, js_receiver_t, uint32_t>;
+using cb_stream_drain_t = js_function_t<void, js_receiver_t>;
+using cb_stream_ack_t = js_function_t<void, js_receiver_t, uint32_t>;
+using cb_stream_send_t = js_function_t<void, js_receiver_t, int32_t, int32_t>;
+using cb_stream_message_t = js_function_t<js_typedarray_span_t<>, js_receiver_t, uint32_t>;
+using cb_stream_realloc_data_t = js_function_t<js_typedarray_span_t<>, js_receiver_t>;
+using cb_stream_realloc_message_t = js_function_t<js_typedarray_span_t<>, js_receiver_t>;
+using cb_stream_close_t = js_function_t<void, js_receiver_t, std::optional<js_object_t>>;
+using cb_stream_firewall_t = js_function_t<uint32_t, js_receiver_t, js_receiver_t, uint32_t, std::string, uint32_t>;
+using cb_stream_remote_changed_t = js_function_t<void, js_receiver_t>;
+
+// udx
+using cb_udx_lookup_t = js_function_t<void, js_receiver_t, std::optional<js_object_t>, std::optional<std::string>, int32_t>;
+
+// interface
+using cb_interface_event_t = js_function_t<void, js_receiver_t>;
+using cb_interface_close_t = js_function_t<void, js_receiver_t>;
+}; // namespace
+
+struct udx_napi_t {
+  udx_t udx;
+
+  uint8_t *read_buf;
+  size_t read_buf_free;
+  // std::span<uint8_t> read_buf; // maybe restore.
+
+  js_deferred_teardown_t *teardown;
+  bool exiting;
+  bool has_teardown;
```

Comment:
```suggestion
```

Initialising `teardown` to `NULL` should make a separate flag unnecessary. We can also handle that in a separate PR seeing how it's also on `main`.